RFC: C implementation of ThingSet device library for Zephyr

b0661 opened the pull request:
## Introduction

It would be nice to have a C language interface ThingSet device library to
support C only applications. 

EDIT: Zephyr module infrastructure will become an extra PR. Commit removed here.

~~As a lot of IoT applications (at least mine !-) are
using the Zepyhr RTOS a Zephyr module would ease the library use.~~

### Problem description

There is (no known) C language interface ThingSet device library for C only
applications. ~~Neither is there Zephyr module support in the ThingSet device library
code base.~~

### Proposed change

Base the ThingSet device library on a C implementation with a C++ interface shim. 
~~Provide it as a Zephyr module~~

 * Enables C applications to use the library.
 * Keeps the interface the same for C++ applications.
 * Reduces maintenance burden as the C and C++ library basically uses the same code.
 * Allows to use the existing C++ test code.

### Proposed change (Detailed)

Move the *.cpp implementation files to C implementations. Keep the functionality, naming,
interface mostly the same as the current C++ implementation. Only the interface
has to get an additional pointer to the ThingSet object it is working on. 

Use the TS/ts_ namespace for the C implementation to avoid name clashes.

Replace printf calls by Zephyr style macros to provide for configurable environment
specific logging and error reporting.

Keep the public interface of the C++ ThingSet class the same. Make the class a header
only shim that uses the C implementation.

~~Create a zephyr sub-directory to contain all the Zephyr specific module configuration
and additional code that is necessary for Zephyr only. Add some libc functions
to this code that the Zephyr minimal libc does (currently) not provide to make
the library usable without newlib.~~

### Dependencies

Existing C++ applications that directly build from source have to use the
C implementation files. There are no more *.cpp files.

### Concerns and Unresolved Questions

None.

## Alternatives

Alternative would be to use the C++ library implementation in C applications. As
it is much easier to integrate C into C++ than vice versa this alternative was not
regarded.

martinjaeger commented:
Wow, thanks a lot for this contribution!

I had the conversion into a pure C library for Zephyr on my ToDo list since a long time already. Really cool.

I'm currently travelling for two weeks, so it might take a few days until I can provide a proper review.

Out of interest: For what sort of application are you using ThingSet and this library?

I was recently thinking about some small changes to the protocol to provide better support for MQTT integration (adding a topic part behind the `#` of the publication messages). And it was probably not a good idea to allow payload in the GET request in the binary mode, as this can't be translated to CoAP or HTTP. I'll let you know once I have properly documented those issues and ideas for improvement. Which mode are you using? The binary or the text mode?

b0661 commented:
> Out of interest: For what sort of application are you using ThingSet and this library?

It is a gateway/ sensor system with special but low cost sensors. In the past I used Firmata on serial to connect the sensors to the gateway. Now I am experimenting to use ThingSet also with the option to use different transports (SPI, CAN). Gateway and sensors are running on Zephyr. Gateway connection is planned to be WebSocket based. 

> Which mode are you using? The binary or the text mode?

My plan is text mode for service/ maintenance access. Binary for regular transmissions. Probably a text mode shell extension for Zephyr will be my next step.

martinjaeger commented on src/thingset.h:
```diff
@@ -323,17 +338,390 @@ typedef struct TsDataNode {
+     * Stores current authentication status (authentication as "normal" user as default)
+     */
+    uint16_t _auth_flags;
+} ts_object_t;
```
In order to be constistent with naming of typedef-ed structs (see also [here](https://github.com/LibreSolar/coding-style)) I suggest following naming:
```C
typedef struct thingset_ctx {
    // ...
} ThingSetCtx;
```
For native Zephyr one would probably use `struct thingset_ctx` then instead of the typedef.

martinjaeger commented on src/thingset.h:
In order to avoid duplication of the comments (and the risk to forget updating them in both locations after changes) I suggest to keep the C++ shim layer simple without any comments and just add one comment at the top of the class definitions to point the user to the C functions for further information.

martinjaeger commented on src/thingset_priv.h:
```diff
@@ -0,0 +1,206 @@
+
+#define DEBUG 0
+
+#define LOG_DBG(...) printf(__VA_ARGS__)
+#define LOG_ERR(...) printf(__VA_ARGS__)
+#define LOG_ALLOC_STR(str) str
```
Very nice that you included those defines for non-Zephyr applications :)

martinjaeger commented on src/thingset_priv.h:
```diff
@@ -0,0 +1,206 @@
+ * Prepares JSMN parser, performs initial check of payload data and calls get/fetch/patch
+ * functions
+ */
+int ts_priv_txt_process(ts_object_t *ts);
```
I suggest to remove the `_priv` in all functions because it makes the code more noisy. I like the idea of the dedicated `thingset_priv.h`, though.

martinjaeger reviewed (commented):
I've tested the code and everything works as expected with existing charge controller firmware. Good job!

However, I had some difficulties with the integration of the Zephyr module change, so I have to look into that a bit more and make sure that the changes in CMakeLists.txt work also with PlatformIO and within the ESP-IDF.

Can you please split out the Zephyr module commit (number 4) into a dedicated PR so that we can get the rest merged more quickly?

Other than that I've only got some small style / naming comments.

b0661 commented:
>> so that we can get the rest merged more quickly?

Do you prefer to have an extra commit for the changes addressing your comments (should make review easier) or shall I combine it all (without the Zephyr commit) into one?

martinjaeger commented:
> > > so that we can get the rest merged more quickly?
> 
> Do you prefer to have an extra commit for the changes addressing your comments (should make review easier) or shall I combine it all (without the Zephyr commit) into one?

You can either amend f7d00559b584b1b473dbfc556a468e98d618d0ae or add a separate commit addressing the comments. Both fine from my side.
Squashing everything into one is not ideal, I think, as it's good to have commits without functional changes separately (exactly as you did).

martinjaeger commented:
Ah, there we go... almost at the same time. I'll review the new comment immediately :)

martinjaeger commented:
Alright, thanks a lot for the updates!

I tested it again and everything works fine, so I'm gonna merge it now.

Just realizing that Travis CI stopped working, so I'll look into that...
